src/Features/CSharp/Portable/AddAccessibilityModifiers/CSharpAddAccessibilityModifiersDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,125 @@
+            SyntaxTreeAnalysisContext context, SyntaxGenerator generator, 
+            CodeStyleOption<AccessibilityModifiersRequired> option, CompilationUnitSyntax compilationUnit)
+        {
+            ProcessMembers(context, generator, option, compilationUnit.Members);
+        }
+
+        private void ProcessMembers(
+            SyntaxTreeAnalysisContext context, SyntaxGenerator generator, 
+            CodeStyleOption<AccessibilityModifiersRequired> option, 
+            SyntaxList<MemberDeclarationSyntax> members)
+        {
+            foreach (var memberDeclaration in members)
+            {
+                ProcessMemberDeclaration(context, generator, option, memberDeclaration);
+            }
+        }
+
+        private void ProcessMemberDeclaration(
+            SyntaxTreeAnalysisContext context, SyntaxGenerator generator,
+            CodeStyleOption<AccessibilityModifiersRequired> option, MemberDeclarationSyntax member)
+        {
+            if (member.IsKind(SyntaxKind.NamespaceDeclaration, out NamespaceDeclarationSyntax namespaceDeclaration))
+            {
+                ProcessMembers(context, generator, option, namespaceDeclaration.Members);
+            }
+
+            // If we have a class or struct, recurse inwards.
+            if (member.IsKind(SyntaxKind.ClassDeclaration, out TypeDeclarationSyntax typeDeclaration) ||
+                member.IsKind(SyntaxKind.StructDeclaration, out typeDeclaration))
+            {
+                ProcessMembers(context, generator, option, typeDeclaration.Members);
+            }
+
+#if false
+            // Add this once we have the language version for C# that supports accessibility
+            // modifiers on interface methods.
+            if (option.Value == AccessibilityModifiersRequired.Always &&
+                member.IsKind(SyntaxKind.InterfaceDeclaration, out typeDeclaration))
+            {
+                // Only recurse into an interface if the user wants accessibility modifiers on 
+                ProcessTypeDeclaration(context, generator, option, typeDeclaration);
+            }
+#endif
+
+            // Have to have a name to report the issue on.
+            var name = member.GetNameToken();
+            if (name.Kind() == SyntaxKind.None)
+            {
+                return;
+            }
+
+            // If they already have accessibility, no need to report anything.
+            var accessibility = generator.GetAccessibility(member);
+            if (accessibility != Accessibility.NotApplicable)
+            {
+                return;
+            }
+
+            // Certain members never have accessibility. Don't bother reporting on them.
+
+            if (member.IsKind(SyntaxKind.DestructorDeclaration))
+            {
+                return;
+            }
+
+            if (member.IsKind(SyntaxKind.ConstructorDeclaration, out ConstructorDeclarationSyntax constructor) &&
+                constructor.Modifiers.Any(SyntaxKind.StaticKeyword))
```

Comment:
Move this static modifier check into [CanHaveAccessibility](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Workspaces/CodeGeneration/CSharpSyntaxGenerator.cs,1411) ?